pkg/kubelet/dockertools/docker.go

```diff
@@ -180,11 +180,30 @@ func matchImageTagOrSHA(inspected dockertypes.ImageInspect, image string) bool {
 	}
 
 	if isDigested {
-		algo := digest.Digest().Algorithm().String()
-		sha := digest.Digest().Hex()
-		// Look specifically for short and long sha(s)
-		if strings.Contains(inspected.ID, algo+":"+sha) {
-			// We found the short or long SHA requested
+		for _, repoDigest := range inspected.RepoDigests {
+			named, err := dockerref.ParseNamed(repoDigest)
+			if err != nil {
+				glog.V(4).Infof("couldn't parse image RepoDigest reference %q: %v", repoDigest, err)
+				continue
+			}
+			if d, isDigested := named.(dockerref.Digested); isDigested {
+				// Allow prefix matches for shorter SHA values
+				if digest.Digest().Algorithm().String() == d.Digest().Algorithm().String() &&
+					strings.HasPrefix(digest.Digest().Hex(), d.Digest().Hex()) {
+					return true
+				}
+			}
+		}
+
+		// process the ID as a digest
+		id, err := dockerdigest.ParseDigest(inspected.ID)
+		if err != nil {
+			glog.V(4).Infof("couldn't parse image ID reference %q: %v", id, err)
+			return false
+		}
+		// Allow prefix matches for shorter SHA values
+		if digest.Digest().Algorithm().String() == id.Algorithm().String() &&
+			strings.HasPrefix(digest.Digest().Hex(), id.Hex()) {
```

Comment:
I can't think of a supported scenario today that being more strict would break (docker rejects, rkt rejects).  I'll remove prefix matching.